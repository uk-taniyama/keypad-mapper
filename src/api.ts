import { findHidDevices } from 'with-hid';

import { getKeyCount } from './types';
import { dbg, splitUInt16LE } from './utils';

import type { ConsumerAction, KeyAction, KeypadInfo, MouseAction } from './types';
import type { HidDeviceQuery, HidIO } from 'with-hid';

// SECTION devices

/**
 * The default query object for finding keypad devices.
 */
export const defaultKeypadQuery: HidDeviceQuery = {
  vendorId: 0x1189,
  productId: 0x8840,
  usagePage: 0xFF00,
  usage: 0x01,
};

/**
 * Finds keypad devices.
 *
 * However, you can specify a custom query if needed.
 */
export function findKeypadDevices(query: HidDeviceQuery = defaultKeypadQuery) {
  return findHidDevices(query);
}

// !SECTION

// SECTION raw I/O

export const REQUEST_SIZE = 65;

/**
 * Sends a request to the HID device.
 */
export async function sendRequest(hid: HidIO, request: number[]) {
  dbg('send:%s', request.map((i) => i.toString(16).padStart(2, '0')).join(' '));
  const req = [...request];
  for (let i = req.length; i < REQUEST_SIZE; i += 1) {
    req.push(0);
  }
  return hid.write(req);
}

/**
 * Receives a response from the HID device.
 *
 * @throws {Error} If no response is received from the device, an error is thrown with the message "No response received."
 */
export async function recvResponse(device: HidIO) {
  const buffer = await device.read(250);
  if (buffer.length === 0) {
    throw new Error('No response received.');
  }
  dbg('recv:%s', buffer.toString('hex').replace(/(00)+$/, '').replace(/(..)/g, '$1 '));
  return buffer;
}

// !SECTION

// SECTION pack/unpack

/**
 * Packs an array of KeyAction objects into a byte array for transmission.
 *
 * @throws {Error} If the number of keys is invalid, an error is thrown.
 */
export function packKeyActrions(keys: KeyAction[]) {
  const len = keys.length;
  if (len < 0 || len > 18) {
    throw new Error(`Invalid size:${len}`);
  }

  const data: number[] = [];
  data.push(len);
  for (let i = 0; i < len; i += 1) {
    data.push(keys[i].modKey);
    data.push(keys[i].keyCode);
  }

  return data;
}

/**
 * Unpacks a byte buffer into an array of KeyAction objects.
 */
export function unpackKeyActions(buffer: Buffer) {
  //        0  1  2  3  4  5  6  7  8  9  A  B  C  D  E  F 10
  // Key-1 03 fa 0f 01-01-00 00 00 00 00:01:00 12
  // Key-2 03 fa 0f 01-01-00 00 00 00 00:02:00 12:00 13
  const len = buffer.readInt8(0x0A);
  const keys: KeyAction[] = [];
  for (let i = 0; i < len; i += 1) {
    const modKey = buffer.readUInt8(0x0B + 2 * i);
    const keyCode = buffer.readUInt8(0x0B + 2 * i + 1);
    keys.push({ modKey, keyCode });
  }
  return keys;
}

/**
 * Packs a ConsumerAction into a byte array for transmission.
 */
export function packConsumerAction({ id }: ConsumerAction) {
  return [0x01, ...splitUInt16LE(id)];
}

/**
 * Unpacks a byte buffer into a ConsumerAction.
 */
export function unpackConsumerAction(buffer: Buffer) {
  //          0  1  2  3  4  5  6  7  8  9  A  B  C
  // AL-Calc 03 fa 02 02-02-00 00 00 00 00:01:92 01
  return { id: buffer.readUInt16LE(0x0B) };
}

// Modified Keys States     1
// Button States            1 Button press states (e.g., left, right)
// X-Axis Movement          1 Horizontal movement (relative)
// Y-Axis Movement          1 Vertical movement (relative)
// Wheel Movement           1 Wheel rotation (vertical scroll)
// Horizontal Wheel (Tilt)  1 Tilt wheel movement (horizontal scroll)

/**
 * Packs a MouseAction object into a byte array.
 */
export function packMouseAction(action: MouseAction) {
  return [0x04, action.modKey, action.button, action.x, action.y, action.wheel];
}

/**
 * Unpacks a byte buffer into a MouseAction object.
 */
export function unpackMouseAction(buffer: Buffer): MouseAction {
  //        0  1  2  3  4  5  6  7  8  9  A  B  C  D  E  F 10
  // L-BTN 03 fa 01 02 03 00 00 00 00 00:04:00 01:00 00 00 00
  // M-BTN 03 fa 02 02 03 00 00 00 00 00:04:00 04:00 00 00 00
  // R-BTN 03 fa 03 02 03 00 00 00 00 00:04:00 02:00 00 00 00
  //   WH+ 03 fa 04 02 03 00 00 00 00 00:04:00.00:00 00 01 00
  //   WH- 03 fa 05 02 03 00 00 00 00 00:04:00.00:00 00 ff 00
  // C-WH+ 03 fa 06 02 03 00 00 00 00 00:04:01.00:00 00 01 00
  // C-WH- 03 fa 07 02 03 00 00 00 00 00:04:01.00:00 00 ff 00
  // S-WH+ 03 fa 08 02 03 00 00 00 00 00:04:02.00:00 00 01 00
  // S-WH- 03 fa 09 02 03 00 00 00 00 00:04:02.00:00 00 ff 00
  // A-WH+ 03 fa 0a 02 03 00 00 00 00 00:04:04.00:00 00 01 00
  // A-WH- 03 fa 0b 02 03 00 00 00 00 00:04:04.00:00 00 ff 00
  const modKey = buffer.readUInt8(0x0B);
  const button = buffer.readUInt8(0x0C);
  const x = buffer.readUInt8(0x0D);
  const y = buffer.readUInt8(0x0E);
  const wheel = buffer.readUInt8(0x0F);
  return {
    modKey, button, x, y, wheel,
  };
}

/**
 * Base interface for KeyMap types.
 */
export interface KeyMapBase {
  type: number;
}

/**
 * Represents an unknown KeyMap with raw buffer data.
 */
export interface KeyMapUnknown extends KeyMapBase {
  buffer: Buffer;
}

// KeyMap Types as constants
export const KeyMapTypes = {
  keys: 0x01,
  consumer: 0x02,
  mouse: 0x03,
} as const;

/**
 * Represents a KeyMap that contains a list of KeyAction objects triggered by key presses.
 */
export interface KeyMapKeys extends KeyMapBase {
  type: typeof KeyMapTypes['keys'];
  keys: KeyAction[];
}

export function isKeyMapKeys(keyMap: KeyMapBase): keyMap is KeyMapKeys {
  return keyMap.type === KeyMapTypes.keys;
}

/**
 * Creates a KeyMapKeys object from an array of KeyAction objects.
 */
export function createKeyMapKeys(keys: KeyAction[]): KeyMapKeys {
  return { type: KeyMapTypes.keys, keys };
}

/**
 * Represents a KeyMap that contains a consumer device action triggered by key presses.
 */
export interface KeyMapConsumer extends KeyMapBase {
  type: typeof KeyMapTypes['consumer'];
  consumer: ConsumerAction;
}

export function isKeyMapConsumer(keyMap: KeyMapBase): keyMap is KeyMapConsumer {
  return keyMap.type === KeyMapTypes.consumer;
}

/**
 * Creates a KeyMapMouse object from a MouseAction.
 */
export function createKeyMapConsumer(action: ConsumerAction): KeyMapConsumer {
  return { type: KeyMapTypes.consumer, consumer: action };
}

/**
 * Represents a KeyMap that contains a mouse action triggered by key presses.
 */
export interface KeyMapMouse extends KeyMapBase {
  type: typeof KeyMapTypes['mouse'];
  mouse: MouseAction;
}

export function isKeyMapMouse(keyMap: KeyMapBase): keyMap is KeyMapMouse {
  return keyMap.type === KeyMapTypes.mouse;
}

/**
 * Creates a KeyMapMouse object from a MouseAction.
 */
export function createKeyMapMouse(action: MouseAction): KeyMapMouse {
  return { type: KeyMapTypes.mouse, mouse: action };
}

export type KeyMap = KeyMapUnknown | KeyMapKeys | KeyMapConsumer | KeyMapMouse;
export type KeyMapList = KeyMap[];
export type KeyMapTable = KeyMapList[];

/**
 * Packs a KeyMap into a byte array.
 *
 * @throws {Error} Throws an error if the KeyMap type is unsupported.
 */
export function packKeyMap(keyMap: KeyMap) {
  let data: number[];
  if (isKeyMapKeys(keyMap)) {
    data = packKeyActrions(keyMap.keys);
  } else if (isKeyMapConsumer(keyMap)) {
    data = packConsumerAction(keyMap.consumer);
  } else if (isKeyMapMouse(keyMap)) {
    data = packMouseAction(keyMap.mouse);
  } else {
    throw new Error(`Unsupported KeyMap type: ${keyMap.type}.`);
  }

  //             type  ?  ?  ?  ?  ?
  return [keyMap.type, 0, 0, 0, 0, 0, ...data];
}

export interface KeyMapWithId {
  layerId: number;
  keyId: number;
  keyMap: KeyMap;
}

/**
 * Unpacks a KeyMap with its associated layerId and keyId from a buffer.
 *
 * @throws {Error} Throws an error if the KeyMap type is unsupported.
 */
export function unpackKeyMapWithId(buffer: Buffer): KeyMapWithId {
  //    0  1  2  3  4  5  6  7  8  9  A  B  C  D  E  F 10
  //    3 fa  k  l  t  ?  ?  ?  ?  ?
  // K 03 fa 0f 01-01 00 00 00 00 00 01 00 12
  // M 03 fa 10 01-03 00 00 00 00 00 04 00 00 00 00 ff 00
  // C 03 fa 01 02 02 00 00 00 00 00 01 b5 00
  const keyId = buffer.readUInt8(2);
  const layerId = buffer.readUInt8(3);
  const type = buffer.readUInt8(4);
  switch (type) {
    case KeyMapTypes.keys:
      return {
        layerId, keyId, keyMap: { type, keys: unpackKeyActions(buffer) },
      };
    case KeyMapTypes.consumer:
      return {
        layerId, keyId, keyMap: { type, consumer: unpackConsumerAction(buffer) },
      };
    case KeyMapTypes.mouse:
      return {
        layerId, keyId, keyMap: { type, mouse: unpackMouseAction(buffer) },
      };
    default:
      return {
        layerId, keyId, keyMap: { type, buffer },
      };
  }
}

// !SECTION

// SECTION api

// TODO layer数はどうやって知れるのかはナゾ

export const packetReadKeypadInfo = [0x03, 0xfb, 0xfb, 0xfb];
export const packetEnd = [0x03, 0xfd, 0xfe, 0xff];

/**
 * Reads keypad information from a HID device.
 */
export async function readKeypadInfo(hid: HidIO) {
  await sendRequest(hid, packetReadKeypadInfo);

  const res = await recvResponse(hid);

  const result: KeypadInfo = {
    keys: res.readUInt8(2),
    knobs: res.readUInt8(3),
  };

  return result;
}

export interface ReadKeypadKeyMapsRequest {
  keys: number;
  knobs: number;
  layerId: number;
}

/**
 * Packs the parameters into a request packet for readKeypadKeyMaps.
 */
export function packReadKeypadKeyMaps({ keys, knobs, layerId }: ReadKeypadKeyMapsRequest) {
  return [0x03, 0xfa, keys, knobs, layerId];
}

/**
 * Reads the keymap configuration for a kaypad from a HID device.
 *
 * @throws {Error} Throws an error if the layer ID in the response does not match the requested layer.
 */
export async function readKeypadKeyMaps(hid: HidIO, param: ReadKeypadKeyMapsRequest) {
  await sendRequest(hid, packReadKeypadKeyMaps(param));

  const { layerId, ...info } = param;

  const result: KeyMapList = [];

  const count = getKeyCount(info);
  for (let i = 0; i < count; i += 1) {
    // eslint-disable-next-line no-await-in-loop
    const buffer = await recvResponse(hid);
    const keyMapWithId = unpackKeyMapWithId(buffer);
    if (keyMapWithId.layerId !== layerId) {
      throw new Error(`Invalid layer response: ${buffer}`);
    }
    result[keyMapWithId.keyId - 1] = keyMapWithId.keyMap;
  }
  return result;
}

/**
 * Reads the keymap table for a kaypad from a HID device.
 *
 * @throws {Error} Throws an error if any layer response is invalid or cannot be read.
 */
export async function readKeypadKeyMapTable(hid: HidIO, info: KeypadInfo, layers: number) {
  const result: KeyMapTable = [];
  for (let i = 0; i < layers; i += 1) {
    // eslint-disable-next-line no-await-in-loop
    const list = await readKeypadKeyMaps(hid, { ...info, layerId: i + 1 });
    result.push(list);
  }
  return result;
}

export interface WriteKeypadKeyMapParam {
  keyId: number;
  layerId: number;
  data: number[];
}

/**
 * Packs the parameters into a request packet for writeKeypadKeyMapRaw.
 */
export function packWriteKeypadKeyMapParam({ keyId, layerId, data }: WriteKeypadKeyMapParam) {
  return [0x03, 0xfd, keyId, layerId, ...data];
}

/**
 * Writes a raw keymap data to the HID device.
 */
export async function writeKeypadKeyMapRaw(hid: HidIO, param: WriteKeypadKeyMapParam) {
  const request = packWriteKeypadKeyMapParam(param);
  return sendRequest(hid, request);
}

/**
 * Writes a keymap to the HID device.
 */
export async function writeKeypadKeyMap(hid: HidIO, { keyId, layerId, keyMap }: KeyMapWithId) {
  return writeKeypadKeyMapRaw(hid, { keyId, layerId, data: packKeyMap(keyMap) });
}

export interface WriteKeypadLedParam {
  layerId: number;
  ledModeId: number;
  ledColorId: number;
}

/**
 * Packs the parameters into a request packet for writeKeypadLed.
 */
export function packWriteKeypadLedParam({ layerId, ledModeId, ledColorId }: WriteKeypadLedParam) {
  //           0  1  2  3  4  5  6  7  8  9  A  B  C
  // 0 RED    03 fe b0 01 08 00 00 00 00 00 01 00 10 : 03 fd fe ff
  // 1 RED    03 fe b0 01 08 00 00 00 00 00 01 00 11 : 03 fd fe ff
  // 2 RED    03 fe b0 01 08 00 00 00 00 00 01 00 12 : 03 fd fe ff
  // 3 ORANGE 03 fe b0 01 08 00 00 00 00 00 01 00 23 : 03 fd fe ff
  // 4 YELLOW 03 fe b0 01 08 00 00 00 00 00 01 00 34 : 03 fd fe ff
  // 0x0A      Layer ID
  // 0x0C-High LED Color ID: Red Orrange Yellow Green Cyan Blue Purple
  // 0x0C-Low  LED Mod ID: 0:None 1:All 2:LT->RB 3:RB->LT 4:Single 5:White
  const ledId = (ledColorId * 16) + ledModeId;
  return [0x03, 0xfe, 0xb0, 0x01, 0x08, 0x00, 0x00, 0x00, 0x00, 0x00, layerId, 0x00, ledId];
}

/**
 * Writes LED settings to the HID device for a specified layer.
 */
export async function writeKeypadLed(hid: HidIO, param: WriteKeypadLedParam) {
  await sendRequest(hid, packWriteKeypadLedParam(param));
  await sendRequest(hid, packetEnd);
}

export interface WriteKeypadDelayTimeParam {
  delayTimeMs: number;
}

/**
 * Packs the parameters into a request packet for writeKeypadDelayTime.
 */
export function packWriteKeypadDelayTimeParam({ delayTimeMs }: WriteKeypadDelayTimeParam) {
  //           0  1  2  3  4  5  6  7  8  9  A  B  C
  // 0 RED    03 fe b0 01 08 00 00 00 00 00 01 00 10 : 03 fd fe ff
  // 1 RED    03 fe b0 01 08 00 00 00 00 00 01 00 11 : 03 fd fe ff
  // 2 RED    03 fe b0 01 08 00 00 00 00 00 01 00 12 : 03 fd fe ff
  // 3 ORANGE 03 fe b0 01 08 00 00 00 00 00 01 00 23 : 03 fd fe ff
  // 4 YELLOW 03 fe b0 01 08 00 00 00 00 00 01 00 34 : 03 fd fe ff
  // 0x0A      Layer ID
  // 0x0C-High LED Color ID: Red Orrange Yellow Green Cyan Blue Purple
  // 0x0C-Low  LED Mod ID: 0:None 1:All 2:LT->RB 3:RB->LT 4:Single 5:White
  //           0  1  2  3  4  5  6  7  8  9  A  B  C
  // 0        03 fe b0 01 05 00 00 00 00 00 01 00 24
  // 10       03 fe b0 01 05 0a 00 00 00 00 01 00 24
  const timeBytes = splitUInt16LE(delayTimeMs);
  return [0x03, 0xfe, 0xb0, 0x01, 0x05, ...timeBytes, 0x00, 0x00, 0x00, 0x01, 0x00, 0x24];
}

/**
 * Writes a delay time setting to the HID device.
 */
export async function writeKeypadDelayTime(hid: HidIO, param: WriteKeypadDelayTimeParam) {
  await sendRequest(hid, packWriteKeypadDelayTimeParam(param));
  await sendRequest(hid, packetEnd);
}

// !SECTION
