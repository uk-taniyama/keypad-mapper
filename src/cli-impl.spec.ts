import { findHidDevices } from 'with-hid';

import { createKeyMapConsumer, createKeyMapKeys, createKeyMapMouse } from './api';
import { ConfigurationError, createKeyActionsFromText, createKeyMapFromActionDef, createKeyMapFromText, createKeyMaps, createParseArgInt, detectKeypad, newCommand, newOption, newOptionInt, normalizeKeypadInfo, parseActionFromText, parseKeyMapFromKeyText, resolveKeyId, runCommand, setupCommonOptions, setupSetKeyMapOptions } from './cli-impl';
import { key } from './key-action';
import { setupSyncMock } from './tests/utils';
import { Ctrl, KnobActions, Shift, media, mouse } from './types';

import type { KeyMap } from './api';
import type { ActionDef } from './cli-impl';
import type { Command } from 'commander';
import type { HidDevice } from 'with-hid';

describe('keyMap from keyDef', () => {
  describe('createKeyActionsFromText', () => {
    it('Error', () => {
      expect.hasAssertions();
      try {
        createKeyActionsFromText('日本語はないので落ちる');
      } catch (e) {
        expect(e).toBeInstanceOf(ConfigurationError);
        expect(`${e}`).toMatchSnapshot();
      }
    });
  });

  describe('createKeyMapFromActionDef', () => {
    it.each<[ActionDef, KeyMap]>([
      ['Hello!', createKeyMapKeys([Shift(key.H), key.E, key.L, key.L, key.O, key.Excl])],
      [key.A, createKeyMapKeys([key.A])],
      [Ctrl(key.A), createKeyMapKeys([Ctrl(key.A)])],
      [[key.A, key.B], createKeyMapKeys([key.A, key.B])],
      [media.MyComputer, createKeyMapConsumer(media.MyComputer)],
      [mouse.LButton, createKeyMapMouse(mouse.LButton)],
      [createKeyMapMouse(mouse.LButton), createKeyMapMouse(mouse.LButton)],
    ])('%s', (def, expected) => {
      const keyMap = createKeyMapFromActionDef(def);
      expect(keyMap).toEqual(expected);
    });
  });

  it('createKeyMaps', () => {
    const keyMaps = createKeyMaps([
      [key.A, null, undefined, key.B],
    ]);
    expect(keyMaps).toEqual([
      { layerId: 1, keyId: 1, keyMap: createKeyMapKeys([key.A]) },
      { layerId: 1, keyId: 4, keyMap: createKeyMapKeys([key.B]) },
    ]);
  });

  describe('parseActionFromText', () => {
    it.each([
      ['A', key.A],
      ['Ctrl+A', Ctrl(key.A)],
      ['LButton', mouse.LButton],
    ])('Success:%s', (text, expected) => {
      const actual = parseActionFromText(text);
      expect(actual).toEqual(expected);
    });

    it.each([
      [''],
      ['Xxx'],
      ['+Ctrl+X'],
      ['Ctrl'],
    ])('Error:%s', (text) => {
      expect.hasAssertions();
      try {
        parseActionFromText(text);
      } catch (e) {
        expect(e).instanceOf(ConfigurationError);
        expect(`${e}`).toMatchSnapshot();
      }
    });
  });

  describe('parseKeyMapFromKeyText', () => {
    it.each([
      ['A', createKeyMapKeys([key.A])],
      ['A B', createKeyMapKeys([key.A, key.B])],
      ['LButton', createKeyMapMouse(mouse.LButton)],
      ['Ctrl+LButton', createKeyMapMouse(Ctrl(mouse.LButton))],
      ['MyComputer', createKeyMapConsumer(media.MyComputer)],
    ])('Success:%s', (text, expected) => {
      const actual = parseKeyMapFromKeyText(text);
      expect(actual).toEqual(expected);
    });

    it.each([
      [''],
      ['A '],
      ['LButton A'],
      ['LButton LButton'],
      ['MyComputer LButton'],
      ['MyComputer MyComputer'],
    ])('Error:%s', (text) => {
      expect.hasAssertions();
      try {
        parseKeyMapFromKeyText(text);
      } catch (e) {
        expect(e).instanceOf(ConfigurationError);
        expect(`${e}`).toMatchSnapshot();
      }
    });
  });

  describe('createKeyMapFromText', () => {
    it.each([
      ['A', createKeyMapKeys([key.A])],
      ['@A', createKeyMapKeys([Shift(key.A)])],
    ])('Success:%s', (text, expected) => {
      const actual = createKeyMapFromText(text);
      expect(actual).toEqual(expected);
    });
  });
});

describe('commander utils', () => {
  it('createParseArgInt', () => {
    const parseArgInt = createParseArgInt('name');
    expect(parseArgInt('10')).toBe(10);
    expect(() => parseArgInt('A')).toThrow(ConfigurationError);
  });

  function parse(command: Command, argv: string[]) {
    const err: string[] = [];
    const out: string[] = [];
    const output = { err, out };
    try {
      command.configureOutput({
        writeErr(str) { err.push(str); },
        writeOut(str) { out.push(str); },
      });
      command.parse(argv, { from: 'user' });
      return { data: command.opts(), output };
    } catch (e) {
      if (e instanceof Error) {
        return { error: e, output };
      }
      return { unknown: e, output };
    }
  }

  describe('newOption', () => {
    it.each([
      ['-x,--xxxx <value>', true, true],
      ['-x,--xxxx [value]', false, false],
    ])('newOption', (flags, mandatory, optionsRequired) => {
      const option = newOption(flags, 'desc', mandatory);
      expect(option.mandatory).toBe(mandatory);
      expect(option.required).toBe(optionsRequired);
      expect(option.envVar).toBe('KEYPAD_XXXX');
    });
  });

  describe('setupCommonOptions', () => {
    it('Success', () => {
      const command = newCommand();
      setupCommonOptions(command);

      const { data } = parse(command, [
        '-P', 'XXXX',
        '-L', '10',
        '-K', '20',
      ]);
      expect(data).toEqual({ path: 'XXXX', layers: 10, keys: 20 });
    });
    it('Error', () => {
      const command = newCommand();
      setupCommonOptions(command);

      const { error } = parse(command, [
        '-L', 'XX',
      ]);
      expect(error?.message).toEqual("option '-L, --layers <value>' argument 'XX' is invalid. It must be a valid integer");
    });
  });

  describe('setupSetKeyMapOptions', () => {
    it('Success', () => {
      const command = newCommand();
      setupSetKeyMapOptions(command);

      const { data } = parse(command, [
        '-P', 'XXXX',
        '-L', '10',
        '-K', '20',
        '-N', '30',
        '-l', '1',
        '-n', '3',
        '-a', 'Left',
      ]);
      expect(data).toEqual({
        path: 'XXXX',
        layers: 10,
        keys: 20,
        knobs: 30,
        layerId: 1,
        knobId: 3,
        knobAction: KnobActions.Left,
      });
    });
    // describe('Error', () => {
    //   it('required', () => {
    //     const command = newCommand();
    //     setupSetKeyMapOptions(command);

    //     const { error } = parse(command, [
    //       '-k', '10',
    //       '-n', '20',
    //     ]);

    //     expect(error?.message).toEqual("error: required option '-l, --layer-id <value>' not specified");
    //   });
    //   it('.conflicts('keyId')', () => {
    //     const command = newCommand();
    //     setupSetKeyMapOptions(command);

    //     const { error } = parse(command, [
    //       '-k', '10',
    //       '-n', '20',
    //       '-l', '10',
    //     ]);
    //     expect(error?.message).toEqual("error: option '-n, --knob-id <value>' cannot be used with option '-k, --key-id <value>'");
    //   });
    // });
  });
});

describe('executeKeypadTask', () => {
  describe('detectKeypad', () => {
    const device: HidDevice = {
      interface: 0, productId: 0, release: 0, vendorId: 0,
    };
    it.each([
      [undefined],
      ['path'],
    ])('Success:%s', (path) => {
      setupSyncMock(findHidDevices, [device]);
      expect(detectKeypad(path)).toEqual(device);
    });

    it.each([
      [0],
      [2],
    ])('Error:%d', (count) => {
      const devices: HidDevice[] = [];
      for (let i = 0; i < count; i += 1) {
        devices.push(device);
      }
      setupSyncMock(findHidDevices, devices);
      expect.hasAssertions();
      try {
        detectKeypad();
      } catch (e) {
        expect(e).instanceOf(ConfigurationError);
        expect(`${e}`).toMatchSnapshot();
      }
    });
  });

  it.each([
    [{}, undefined],
    [{ keys: 1 }, { keys: 1, knobs: 0 }],
    [{ knobs: 2 }, { keys: 0, knobs: 2 }],
    [{ keys: 1, knobs: 2 }, { keys: 1, knobs: 2 }],
  ])('normalizeKeypadInfo:%s', (config, expected) => {
    const actual = normalizeKeypadInfo(config);
    expect(actual).toEqual(expected);
  });

  describe('resolveKeyId', () => {
    it.each([
      [{ keyId: 2 }, 2],
      [{ keyId: 6 }, 6],
      [{ keyId: 2, knobId: 3, knobAction: 'Left' }, 2],
      [{ knobId: 1, knobAction: 'Left' }, 4],
    ])('Success:%s', (config, expected) => {
      const actual = resolveKeyId({ keys: 3, knobs: 1 }, config);
      expect(actual).toBe(expected);
    });
    it.each([
      [{ keyId: 0 }],
      [{ keyId: 7 }],
      [{}],
      [{ knobId: 1 }],
      [{ knobAction: 'Left' }],
      [{ knobId: 1, knobAction: 'Unknown' }],
      [{ knobId: 0, knobAction: 'Left' }],
      [{ knobId: 2, knobAction: 'Left' }],
    ])('Error:%s', (config) => {
      expect.hasAssertions();
      try {
        resolveKeyId({ keys: 3, knobs: 1 }, config);
      } catch (e) {
        expect(e).instanceOf(ConfigurationError);
        expect(`${e}`).toMatchSnapshot();
      }
    });
  });

  describe('runCommand', () => {
    const processExit = process.exit;
    const consoleError = console.error;
    afterEach(() => {
      process.exit = processExit;
      console.error = consoleError;
    });

    function testRunCommand(
      setup: (command: Command) => Command,
      argv: string[],
      run?: (...args: any[]) => Promise<void>,
    ) {
      return new Promise<{
        exit?: true;
        code?: any;
        args?: any[];
        err: string;
        out: string;
      }>((resolve) => {
        const out: string[] = [];
        const err: string[] = [];

        function doResolve(obj: Object) {
          resolve({ ...obj, out: out.join('\n'), err: err.join('\n') });
        }
        const exit = (code: any) => { doResolve({ exit: true, code }); };
        process.exit = exit as any;
        const error = (...args: any) => { err.push(args.join(' ')); };
        console.error = error;
        runCommand([{
          setup(command) {
            command.configureOutput({
              writeOut: (str) => { out.push(str); },
              writeErr: (str) => { err.push(str); },
            });
            return setup(command);
          },
          async run(...args: any[]) {
            args.pop();
            if (run) {
              await run(args);
            }
            doResolve({ args });
          },
        }], argv);
      });
    }

    it('simple', async () => {
      const result = await testRunCommand((command) => {
        return command.command('simple');
      }, ['node', '.js', 'simple']);
      expect(result).toEqual({ args: [{}], err: '', out: '' });
    });

    it('required option. show usage', async () => {
      const result = await testRunCommand((command) => {
        return command.command('simple')
          .addOption(newOptionInt('-i', 'int', true));
      }, ['node', '.js', 'simple']);
      expect(result).toEqual({
        exit: true, code: 1, err: expect.anything(), out: '',
      });
      expect(result.err).toMatch(/error: required option /);
      expect(result.err).toMatch(/Usage/);
      expect(result).toMatchSnapshot();
    });

    it('parse error. show usage', async () => {
      const result = await testRunCommand((command) => {
        return command.command('simple')
          .addOption(newOptionInt('-i <value>', 'int', true));
      }, ['node', '.js', 'simple', '-i', 'xxx']);
      expect(result).toEqual({
        exit: true, code: 1, err: expect.anything(), out: '',
      });
      expect(result.err).toMatch(/error: option /);
      expect(result.err).toMatch(/Usage/);
      expect(result).toMatchSnapshot();
    });

    it('unknown error. message only', async () => {
      const result = await testRunCommand((command) => {
        return command.command('simple');
      }, ['node', '.js', 'simple'], () => { throw new Error('Unkown'); });
      expect(result).toEqual({
        exit: true, code: 1, err: 'Error: Unkown', out: '',
      });
    });
  });
});
