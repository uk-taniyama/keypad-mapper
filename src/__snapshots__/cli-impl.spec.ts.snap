// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`executeKeypadTask > detectKeypad > Error:0 1`] = `"ConfigurationError: cannot find the keypad device"`;

exports[`executeKeypadTask > detectKeypad > Error:2 1`] = `"ConfigurationError: cannot detect the keypad device. Use the '--path' option to specify its path"`;

exports[`executeKeypadTask > resolveKeyId > Error:{ keyId: +0 } 1`] = `"ConfigurationError: invalid key ID: 0. It must be between 1 and 6"`;

exports[`executeKeypadTask > resolveKeyId > Error:{ keyId: 7 } 1`] = `"ConfigurationError: invalid key ID: 7. It must be between 1 and 6"`;

exports[`executeKeypadTask > resolveKeyId > Error:{ knobAction: 'Left' } 1`] = `"ConfigurationError: cannot resolve the key ID. Specify either '--key-id' or both '--knob-id' and '--knob-action'"`;

exports[`executeKeypadTask > resolveKeyId > Error:{ knobId: +0, knobAction: 'Left' } 1`] = `"ConfigurationError: invalid knob ID: 0. It must be between 1 and 1"`;

exports[`executeKeypadTask > resolveKeyId > Error:{ knobId: 1 } 1`] = `"ConfigurationError: cannot resolve the key ID. Specify either '--key-id' or both '--knob-id' and '--knob-action'"`;

exports[`executeKeypadTask > resolveKeyId > Error:{ knobId: 1, knobAction: 'Unknown' } 1`] = `"ConfigurationError: cannot process the knob action: 'Unknown' is invalid"`;

exports[`executeKeypadTask > resolveKeyId > Error:{ knobId: 2, knobAction: 'Left' } 1`] = `"ConfigurationError: invalid knob ID: 2. It must be between 1 and 1"`;

exports[`executeKeypadTask > resolveKeyId > Error:{} 1`] = `"ConfigurationError: cannot resolve the key ID. Specify either '--key-id' or both '--knob-id' and '--knob-action'"`;

exports[`executeKeypadTask > runCommand > parse error. show usage 1`] = `
{
  "code": 1,
  "err": "error: option '-i <value>' argument 'xxx' is invalid. It must be a valid integer

Usage: keypad-mapper simple [options]

Options:
  -i <value>  int (env: KEYPAD_I)
  -h, --help  display help for command
",
  "exit": true,
  "out": "",
}
`;

exports[`executeKeypadTask > runCommand > required option. show usage 1`] = `
{
  "code": 1,
  "err": "error: required option '-i' not specified



Usage: keypad-mapper simple [options]

Options:
  -i          int (env: KEYPAD_I)
  -h, --help  display help for command
",
  "exit": true,
  "out": "",
}
`;

exports[`keyMap from keyDef > createKeyActionsFromText > Error 1`] = `"ConfigurationError: cannot convert character '日' at index 0 into a valid KeyAction. Full text: 日本語はないので落ちる"`;

exports[`keyMap from keyDef > parseActionFromText > Error: 1`] = `"ConfigurationError: cannot process the action string. The string cannot be empty"`;

exports[`keyMap from keyDef > parseActionFromText > Error:+Ctrl+X 1`] = `"ConfigurationError: cannot parse modifier keys from: '+Ctrl+X'"`;

exports[`keyMap from keyDef > parseActionFromText > Error:Ctrl 1`] = `"ConfigurationError: cannot resolve the action: 'Ctrl' is unknown"`;

exports[`keyMap from keyDef > parseActionFromText > Error:Xxx 1`] = `"ConfigurationError: cannot resolve the action: 'Xxx' is unknown"`;

exports[`keyMap from keyDef > parseKeyMapFromKeyText > Error: 1`] = `"ConfigurationError: cannot process the action string. The string cannot be empty"`;

exports[`keyMap from keyDef > parseKeyMapFromKeyText > Error:A  1`] = `"ConfigurationError: cannot process the action string. The string cannot be empty"`;

exports[`keyMap from keyDef > parseKeyMapFromKeyText > Error:LButton A 1`] = `"ConfigurationError: cannot set conflicting actions"`;

exports[`keyMap from keyDef > parseKeyMapFromKeyText > Error:LButton LButton 1`] = `"ConfigurationError: cannot process the key map. Only one mouse action is allowed"`;

exports[`keyMap from keyDef > parseKeyMapFromKeyText > Error:MyComputer LButton 1`] = `"ConfigurationError: cannot set conflicting actions"`;

exports[`keyMap from keyDef > parseKeyMapFromKeyText > Error:MyComputer MyComputer 1`] = `"ConfigurationError: cannot process the key map. Only one consumer action is allowed"`;
